File: bindings/dart/test/metadata_test.dart
import 'package:quiver_db/quiver_db.dart';
import 'package:quiver_db/src/ffi/bindings.dart';
import 'package:test/test.dart';
import 'package:path/path.dart' as path;

void main() {
  // Path to central tests folder
  final testsPath = path.join(
    path.current,
    '..',
    '..',
    'tests',
  );

  group('Scalar Metadata', () {
    test('returns metadata for text attribute', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final metadata = db.getScalarMetadata('Collection', 'label');
        expect(metadata.name, equals('label'));
        expect(metadata.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_STRING));
        expect(metadata.notNull, isTrue);
        expect(metadata.primaryKey, isFalse);
      } finally {
        db.close();
      }
    });

    test('returns metadata for nullable integer attribute', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final metadata = db.getScalarMetadata('Collection', 'some_integer');
        expect(metadata.name, equals('some_integer'));
        expect(metadata.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_INTEGER));
        expect(metadata.notNull, isFalse);
        expect(metadata.primaryKey, isFalse);
      } finally {
        db.close();
      }
    });

    test('returns metadata for nullable real attribute', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final metadata = db.getScalarMetadata('Collection', 'some_float');
        expect(metadata.name, equals('some_float'));
        expect(metadata.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_FLOAT));
        expect(metadata.notNull, isFalse);
        expect(metadata.primaryKey, isFalse);
      } finally {
        db.close();
      }
    });

    test('returns metadata for primary key', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final metadata = db.getScalarMetadata('Collection', 'id');
        expect(metadata.name, equals('id'));
        expect(metadata.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_INTEGER));
        expect(metadata.primaryKey, isTrue);
      } finally {
        db.close();
      }
    });

    test('throws on non-existent collection', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        expect(
          () => db.getScalarMetadata('NonExistent', 'label'),
          throwsA(isA<DatabaseException>()),
        );
      } finally {
        db.close();
      }
    });

    test('throws on non-existent attribute', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        expect(
          () => db.getScalarMetadata('Collection', 'nonexistent'),
          throwsA(isA<DatabaseException>()),
        );
      } finally {
        db.close();
      }
    });
  });

  group('Vector Metadata', () {
    test('returns all attributes in vector group', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final metadata = db.getVectorMetadata('Collection', 'values');
        expect(metadata.groupName, equals('values'));
        expect(metadata.valueColumns.length, equals(2));

        final colNames = metadata.valueColumns.map((a) => a.name).toList();
        expect(colNames, contains('value_int'));
        expect(colNames, contains('value_float'));

        final valueIntCol = metadata.valueColumns.firstWhere((a) => a.name == 'value_int');
        expect(valueIntCol.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_INTEGER));

        final valueFloatCol = metadata.valueColumns.firstWhere((a) => a.name == 'value_float');
        expect(valueFloatCol.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_FLOAT));
      } finally {
        db.close();
      }
    });

    test('throws on non-existent vector group', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        expect(
          () => db.getVectorMetadata('Collection', 'nonexistent'),
          throwsA(isA<DatabaseException>()),
        );
      } finally {
        db.close();
      }
    });
  });

  group('Set Metadata', () {
    test('returns all attributes in set group', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final metadata = db.getSetMetadata('Collection', 'tags');
        expect(metadata.groupName, equals('tags'));
        expect(metadata.valueColumns.length, equals(1));

        final tagCol = metadata.valueColumns[0];
        expect(tagCol.name, equals('tag'));
        expect(tagCol.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_STRING));
      } finally {
        db.close();
      }
    });

    test('throws on non-existent set group', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        expect(
          () => db.getSetMetadata('Collection', 'nonexistent'),
          throwsA(isA<DatabaseException>()),
        );
      } finally {
        db.close();
      }
    });
  });

  group('Foreign Key Metadata', () {
    test('returns FK info for foreign key attribute', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'relations.sql'),
      );
      try {
        final metadata = db.getScalarMetadata('Child', 'parent_id');
        expect(metadata.isForeignKey, isTrue);
        expect(metadata.referencesCollection, equals('Parent'));
        expect(metadata.referencesColumn, equals('id'));
      } finally {
        db.close();
      }
    });

    test('returns FK info for self-referencing key', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'relations.sql'),
      );
      try {
        final metadata = db.getScalarMetadata('Child', 'sibling_id');
        expect(metadata.isForeignKey, isTrue);
        expect(metadata.referencesCollection, equals('Child'));
        expect(metadata.referencesColumn, equals('id'));
      } finally {
        db.close();
      }
    });

    test('returns no FK info for non-FK attribute', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'relations.sql'),
      );
      try {
        final metadata = db.getScalarMetadata('Child', 'label');
        expect(metadata.isForeignKey, isFalse);
        expect(metadata.referencesCollection, isNull);
        expect(metadata.referencesColumn, isNull);
      } finally {
        db.close();
      }
    });

    test('list_scalar_attributes includes FK info', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'relations.sql'),
      );
      try {
        final attributes = db.listScalarAttributes('Child');
        final parentAttr = attributes.firstWhere((a) => a.name == 'parent_id');
        expect(parentAttr.isForeignKey, isTrue);
        expect(parentAttr.referencesCollection, equals('Parent'));

        final labelAttr = attributes.firstWhere((a) => a.name == 'label');
        expect(labelAttr.isForeignKey, isFalse);
        expect(labelAttr.referencesCollection, isNull);
      } finally {
        db.close();
      }
    });
  });

  group('List Scalar Attributes', () {
    test('returns all scalar attributes with metadata', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final attributes = db.listScalarAttributes('Collection');
        final attrNames = attributes.map((a) => a.name).toList();
        expect(attrNames, contains('id'));
        expect(attrNames, contains('label'));
        expect(attrNames, contains('some_integer'));
        expect(attrNames, contains('some_float'));

        // Verify metadata is included
        final labelAttr = attributes.firstWhere((a) => a.name == 'label');
        expect(labelAttr.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_STRING));
        expect(labelAttr.notNull, isTrue);

        final someIntAttr = attributes.firstWhere((a) => a.name == 'some_integer');
        expect(someIntAttr.dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_INTEGER));
        expect(someIntAttr.notNull, isFalse);
      } finally {
        db.close();
      }
    });
  });

  group('List Vector Groups', () {
    test('returns all vector groups with metadata', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final groups = db.listVectorGroups('Collection');
        final groupNames = groups.map((g) => g.groupName).toList();
        expect(groupNames, contains('values'));

        // Verify metadata is included
        final valuesGroup = groups.firstWhere((g) => g.groupName == 'values');
        expect(valuesGroup.valueColumns.length, equals(2));
        final colNames = valuesGroup.valueColumns.map((a) => a.name).toList();
        expect(colNames, contains('value_int'));
        expect(colNames, contains('value_float'));
      } finally {
        db.close();
      }
    });
  });

  group('Read Set Group By ID', () {
    test('reads set group by id', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'all_types.sql'),
      );
      try {
        db.createElement('AllTypes', {'label': 'Item 1'});
        db.updateSetIntegers('AllTypes', 'code', 1, [10, 20, 30]);

        final rows = db.readSetGroupById('AllTypes', 'codes', 1);
        expect(rows, isNotEmpty);
        final codes = rows.map((row) => row['code']).toList()..sort();
        expect(codes, equals([10, 20, 30]));
      } finally {
        db.close();
      }
    });
  });

  group('List Set Groups', () {
    test('returns all set groups with metadata', () {
      final db = Database.fromSchema(
        ':memory:',
        path.join(testsPath, 'schemas', 'valid', 'collections.sql'),
      );
      try {
        final groups = db.listSetGroups('Collection');
        final groupNames = groups.map((g) => g.groupName).toList();
        expect(groupNames, contains('tags'));

        // Verify metadata is included
        final tagsGroup = groups.firstWhere((g) => g.groupName == 'tags');
        expect(tagsGroup.valueColumns.length, equals(1));
        expect(tagsGroup.valueColumns[0].name, equals('tag'));
        expect(tagsGroup.valueColumns[0].dataType, equals(quiver_data_type_t.QUIVER_DATA_TYPE_STRING));
      } finally {
        db.close();
      }
    });
  });
}

File: bindings/js/test/metadata.test.ts
import { assert, assertEquals, assertNotEquals, assertThrows } from "jsr:@std/assert";
const __dirname = import.meta.dirname!;
import { join } from "jsr:@std/path";
import { Database } from "../src/index.ts";
import type { GroupMetadata, ScalarMetadata } from "../src/index.ts";

const SCHEMA_PATH = join(
  __dirname,
  "..",
  "..",
  "..",
  "tests",
  "schemas",
  "valid",
  "collections.sql",
);

Deno.test({ name: "getScalarMetadata", sanitizeResources: false }, async (t) => {
  await t.step("returns correct metadata for an integer scalar", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const meta = db.getScalarMetadata("Collection", "some_integer");
      assertEquals(meta.name, "some_integer");
      assertEquals(typeof meta.dataType, "number");
      assertEquals(meta.notNull, false);
      assertEquals(meta.primaryKey, false);
      assertEquals(meta.defaultValue, null);
      assertEquals(meta.isForeignKey, false);
      assertEquals(meta.referencesCollection, null);
      assertEquals(meta.referencesColumn, null);
    } finally {
      db.close();
    }
  });

  await t.step("returns correct metadata for the primary key", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const meta = db.getScalarMetadata("Collection", "id");
      assertEquals(meta.name, "id");
      assertEquals(meta.primaryKey, true);
      // SQLite INTEGER PRIMARY KEY does not set the not_null PRAGMA flag
      // even though the column is implicitly non-nullable
      assertEquals(typeof meta.notNull, "boolean");
    } finally {
      db.close();
    }
  });

  await t.step("returns correct metadata for a foreign key", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const meta = db.getScalarMetadata("Collection_vector_values", "id");
      assertEquals(meta.isForeignKey, true);
      assertEquals(meta.referencesCollection, "Collection");
      assertEquals(meta.referencesColumn, "id");
    } finally {
      db.close();
    }
  });

  await t.step("throws on nonexistent attribute", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      assertThrows(() => db.getScalarMetadata("Collection", "nonexistent"));
    } finally {
      db.close();
    }
  });
});

Deno.test({ name: "getVectorMetadata / getSetMetadata / getTimeSeriesMetadata", sanitizeResources: false }, async (t) => {
  await t.step("getVectorMetadata returns correct group", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const meta = db.getVectorMetadata("Collection", "values");
      assertEquals(meta.groupName, "values");
      assertEquals(meta.dimensionColumn, null);
      assert(meta.valueColumns.length >= 1);
      for (const col of meta.valueColumns) {
        assertEquals(typeof col.name, "string");
        assertEquals(typeof col.dataType, "number");
      }
    } finally {
      db.close();
    }
  });

  await t.step("getSetMetadata returns correct group", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const meta = db.getSetMetadata("Collection", "tags");
      assertEquals(meta.groupName, "tags");
      assertEquals(meta.dimensionColumn, null);
      assert(meta.valueColumns.length >= 1);
    } finally {
      db.close();
    }
  });

  await t.step("getTimeSeriesMetadata returns non-null dimensionColumn", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const meta = db.getTimeSeriesMetadata("Collection", "data");
      assertEquals(meta.groupName, "data");
      assertEquals(meta.dimensionColumn, "date_time");
      assert(meta.valueColumns.length >= 1);
    } finally {
      db.close();
    }
  });
});

Deno.test({ name: "listScalarAttributes", sanitizeResources: false }, async (t) => {
  await t.step("lists all scalar attributes", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const attrs = db.listScalarAttributes("Collection");
      assertEquals(Array.isArray(attrs), true);
      // Collection has: id, label, some_integer, some_float
      assert(attrs.length >= 3);

      const someInt = attrs.find((a: ScalarMetadata) => a.name === "some_integer");
      assertNotEquals(someInt, undefined);
      assertEquals(typeof someInt!.dataType, "number");
      assertEquals(someInt!.primaryKey, false);
    } finally {
      db.close();
    }
  });
});

Deno.test({ name: "listVectorGroups / listSetGroups / listTimeSeriesGroups", sanitizeResources: false }, async (t) => {
  await t.step("listVectorGroups returns vector groups", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const groups = db.listVectorGroups("Collection");
      assert(groups.length >= 1);

      const values = groups.find((g: GroupMetadata) => g.groupName === "values");
      assertNotEquals(values, undefined);
      assertEquals(values!.dimensionColumn, null);
    } finally {
      db.close();
    }
  });

  await t.step("listSetGroups returns set groups", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const groups = db.listSetGroups("Collection");
      assert(groups.length >= 1);

      const tags = groups.find((g: GroupMetadata) => g.groupName === "tags");
      assertNotEquals(tags, undefined);
      assertEquals(tags!.dimensionColumn, null);
    } finally {
      db.close();
    }
  });

  await t.step("listTimeSeriesGroups returns time series groups with dimensionColumn", () => {
    const db = Database.fromSchema(":memory:", SCHEMA_PATH);
    try {
      const groups = db.listTimeSeriesGroups("Collection");
      assert(groups.length >= 1);

      const data = groups.find((g: GroupMetadata) => g.groupName === "data");
      assertNotEquals(data, undefined);
      assertEquals(data!.dimensionColumn, "date_time");
    } finally {
      db.close();
    }
  });
});
